hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/BaseSparkInternalRowReaderContext.java

```diff
@@ -83,7 +86,26 @@ public Object getValue(InternalRow row, Schema schema, String fieldName) {
 
   @Override
   public String getRecordKey(InternalRow row, Schema schema) {
-    return getFieldValueFromInternalRow(row, schema, RECORD_KEY_METADATA_FIELD).toString();
+    Object key = getFieldValueFromInternalRow(row, schema, RECORD_KEY_METADATA_FIELD);
+    if (key != null) {
+      return key.toString();
+    }
+    return null;
+  }
+
+  @Override
+  public String getRecordKey(InternalRow row, Schema schema, TypedProperties props) {
+    String key = getRecordKey(row, schema);
```

Comment:
can you help me understand why do we need these changes?

Reply:
During my experiments, the record key may be found by checking different configs. Existing code tries to retrieve it in only one way, which will cause null-key issue.

Reply:
In my experiments, the schema may contain metadata fields, but there is no such columns in the data. Therefore, these metadata fields are NULL after retrieved from the disk. When we tried to get the record key, only the `_record_key` is used, which gives us NULL, and cause the merge failure. To avoid such problem, we should try to find the key field; there are different ways to find out the key field, either from the table config or from the write config.